src/v/kafka/server/handlers/fetch.cc

```diff
@@ -139,13 +142,28 @@ static ss::future<read_result> read_from_partition(
                   part.high_watermark());
             }
         }
-
+    } catch (const cloud_storage::download_exception&) {
+        e = std::current_exception();
+        known_exception = true;
+    } catch (const ss::gate_closed_exception&) {
+        e = std::current_exception();
+        known_exception = true;
     } catch (...) {
         e = std::current_exception();
+        known_exception = false;
     }
 
     co_await std::move(rdr.reader).release()->finally();
 
+    if (e && known_exception) {
+        vlog(
+          klog.info,
+          "exception while reading topic_partition: {} exception: {}",
+          part.ntp().tp,
+          e);
+        co_return read_result(error_code::kafka_storage_error, start_o, hw);
```

Comment:
Not sure how big of a deal this is, but `kafka_storage_error` is only a valid response in version 6 and above of fetch response.

cc @graphcareful @dotnwat

Reply:
https://github.com/redpanda-data/redpanda/blob/21b01015f17cbebe40af53717461f86770a53356/src/v/kafka/protocol/schemata/fetch_response.json#L30

Reply:
good point. @ballard26 should we return a more generic error here that isn't associated with a particular minimum version?

Reply:
also, @ballard26 should we wait and merge this after we branch 24.1?

Reply:
> good point. @ballard26 should we return a more generic error here that isn't associated with a particular minimum version?

Sure, any retriable error would work for me. My quick survey of two clients(librdkafka and franzgo) show that they only really treat errors as either retriable or not. I don't see any special handling of specific ones beyond that.

Reply:
> also, @ballard26 should we wait and merge this after we branch 24.1?

Fine with waiting until the branch. It may be worth letting this one stew in dev for awhile given the hesitancy.

Reply:
Do folks have a recommendation for a retriable error? It looks like we like using `not_leader` for a lot of our retriable errors. It'd be nice to may have an error that communicates that the client metadata is correct, but the request failed for other reasons though.

Reply:
i don't have a good suggestion, maybe the storage error is the best one. newer clients would understand it and retry, and older clients would do what? drop the connection? in that case it would be no worse than before? not_leader other retriable errors like that go back far enough for compatibility to not be an issue, but they might also trigger more expensive retry strategies?

Reply:
> Do folks have a recommendation for a retriable error?
Maybe REQUEST_TIMED_OUT?  Can we check the request API version and if it's >=6 return STORAGE_ERROR, and then others some other retriable one?

Reply:
How about [NOT_LEADER_OR_FOLLOWER](https://github.com/apache/kafka/commit/f221a2e59ebb6c0deccb35cbcf92576954803231#diff-1ef70ab50c661cda6afb43edf35d5d0af011b83d6da3731cca73f77756a38d40R554)? Which is `kafka::error_code::not_leader_for_partition`